Release/1.3.0

oguzhanunlu opened the pull request:
* Fix database cursor usage (#330)
* Comply with isLimitAdTrackingEnabled (#329)
* Integrate Snowplow Micro (#318)
* Trim travis setup (#324)
* Remove vagrant setup (#317)

colmsnowplow commented on snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/tracker/Tracker.java:
```diff
@@ -957,4 +964,27 @@ public void removeGlobalContext(String tag) {
+     * Enables gdpr context to be sent with every event
+     */
+    public void enableGDPRContext(BASIS basisForProcessing, String documentId, String documentVersion, String documentDescription) {
+        String basis = basisForProcessing.toString().toLowerCase();
```
Since this is an [enum](https://github.com/snowplow/iglu-central/blob/248b3be05b097693363e014b902008fa6cb3bfc8/schemas/com.snowplowanalytics.snowplow/gdpr/jsonschema/1-0-0#L14) in the schema, we've opted to [make it an enum in the javascript tracker](https://github.com/snowplow/snowplow-javascript-tracker/blob/a2a12a5c1a2065db8df250eb1bf6e31900e81a0c/src/js/tracker.js#L111).

Not sure how practicable that is for Android

oguzhanunlu commented:
Thanks @colmsnowplow ! I've extended the PR after your review by 1 commit, could you please review that last commit too?

jbeemster commented on .travis.yml:
```diff
@@ -9,54 +9,38 @@ dist: trusty
 env:
   global:
-    - EMULATOR_API=24
+    - EMULATOR_API=16
```
How come we are dropping the API back so much?

oguzhanunlu replied:
The tracker supports SDK 14 at the minimum and emulator startup time of 16 is lower than 24 in Travis. It reduced build time from ~30 min to ~22-23 min. Do you think having a recent version will bring an advantage?

jbeemster replied:
So API version 16 is used only by 1.2% of devices according to this page:

> https://developer.android.com/about/dashboards

I would argue that using a later version is more useful in that you are testing on more relevant releases.

oguzhanunlu replied:
I see your point, I'll revert it 👍

jbeemster commented on .travis.yml:
```diff
@@ -9,54 +9,38 @@ dist: trusty
-  - sdkmanager --list --verbose || true
+#  - docker pull snowplow/snowplow-micro:0.1.0
+#  - wget https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip
+#  - unzip ./ngrok-stable-linux-amd64.zip
```
If these are commented out do we need them?

oguzhanunlu replied:
Micro integration is half complete, i.e. integration tests are run locally only, not in Travis, and I opened #328 to address the other half. These lines are commented out because they aren't necessary until 328 is done. Would you prefer to save this incomplete work somewhere else and delete from here? Or anything else?

jbeemster replied:
Maybe add a comment that these commented lines are related to that ticket and leave them in then?

jbeemster commented on .travis.yml:
```diff
@@ -65,6 +49,10 @@ script:
+#  - docker run -d --mount type=bind,source=$TRAVIS_BUILD_DIR/micro_config,destination=/config -p 9090:9090 snowplow/snowplow-micro:0.1.0 --collector-config /config/micro.conf --iglu /config/iglu.json
+#  - docker ps -a
+#  - echo $MICRO_SUBDOMAIN
+#  - ./ngrok http -authtoken $NGROK_AUTH -subdomain $MICRO_SUBDOMAIN -log=stdout 9090 > /dev/null &
```
Same here

jbeemster commented on snowplow-demo-app/build.gradle:
```diff
@@ -38,8 +56,11 @@ dependencies {
+    androidTestImplementation 'androidx.test:rules:1.2.0'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
+}
```
Nit newline

jbeemster commented on snowplow-demo-app/src/androidTest/java/com/snowplowanalytics/snowplowtrackerdemo/uitest/UITest.java:
```diff
@@ -0,0 +1,179 @@
+package com.snowplowanalytics.snowplowtrackerdemo.uitest;
```
Nit header missing

jbeemster commented on snowplow-demo-app/src/androidTest/java/com/snowplowanalytics/snowplowtrackerdemo/uitest/UITest.java:
```diff
@@ -0,0 +1,179 @@
+        onView(withId(R.id.btn_lite_start)).perform(click());
+
+        // TODO this is not best practice for idling resources
+        Thread.sleep(20000);
```
This seems excessive and is introducing 20 seconds to a test suite!  This should really be refactored to not require any sleeps.

oguzhanunlu replied:
It was based on demo app, shorter sleeps can end the tests without letting the app send all events.

I'm aware of the disadvantage and I was planning to change the approach later since this isn't a proper way. Although test duration won't change much, I'll change the approach right in this release.

jbeemster replied:
I think for this kind of thing its better not to leave it!  Let me know if you need any help solving it.

jbeemster commented on snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/tracker/Tracker.java:
```diff
@@ -957,4 +964,27 @@ public void removeGlobalContext(String tag) {
+    /**
+     * Enables gdpr context to be sent with every event
+     */
+    public void enableGDPRContext(BASIS basisForProcessing, String documentId, String documentVersion, String documentDescription) {
```
This function seems like it should only set the global values if `gdprContext = false`.  Otherwise if you are mid sending events when it is on and switch it off you might get a thread race condition.

Also worth making this function thread safe with tracking calls to ensure you cannot have any race conditions between these tracker variables.

Atomic booleans could be useful here!

oguzhanunlu replied:
> This function seems like it should only set the global values if gdprContext = false. 

It mimics how our js tracker offers gdpr context. Are you sure on this? 

Thanks for catching thread-safety. I'll make the method `synchronized`.

jbeemster replied:
Mimic the same behaviour then but thread safety is really important for Tracker functions!

oguzhanunlu commented:
Hey @jbeemster , I addressed all your feedback, especially on `thread.sleep` usage in UI test and thread safety of gdpr context mechanisms, however the scope of the release has to be cut (configurable emitter post path and gdpr context is postponed) and I need to ask for review again on updated commits. Thanks in advance.
